packages/git/src/browser/dirty-diff/dirty-diff-manager.ts

```diff
@@ -0,0 +1,240 @@
+            this.models.set(uri, model);
+            editorWidget.disposed.connect(() => {
+                model.editorCounter--;
+                if (model.editorCounter < 1) {
+                    this.models.delete(uri);
+                    model.dispose();
+                }
+            });
+        }
+    }
+
+    protected async createNewModel(editor: TextEditor): Promise<DirtyDiffModel> {
+        const uri = editor.document.uri;
+        const model = new DirtyDiffModel(uri, u => this.readGitResourceContents(u));
+        model.editorCounter++;
+        const gitStatus = this.repositoryTracker.selectedRepositoryStatus;
+        if (gitStatus) {
+            const changes = gitStatus.changes.filter(c => c.uri === uri);
+            await model.handleGitStatusUpdate(changes);
+        }
+        model.handleDocumentChanged(editor.document.getText());
+        model.onDityDiffUpdate(e => this.onDityDiffUpdateEmitter.fire(e));
+        model.enabled = this.isEnabled();
+        return model;
+    }
+
+    protected async readGitResourceContents(uri: URI): Promise<string> {
+        const gitResource = await this.gitResourceResolver.getResource(uri);
+        return gitResource.readContents();
+    }
+
+    protected handleDocumentChanged(uri: string): void {
+        const existingModel = this.models.get(uri);
+        if (existingModel) {
+            const document = this.workspace.textDocuments.find(d => d.uri === uri);
+            if (document) {
+                existingModel.handleDocumentChanged(document.getText());
+            }
+        }
+    }
+
+    protected async handleGitStatusUpdate(status: WorkingDirectoryStatus): Promise<void> {
+        const uris = new Set(this.models.keys());
+        const relevantChanges = status.changes.filter(c => uris.has(c.uri));
+        for (const uri of uris) {
+            const model = this.models.get(uri)!;
+            const changes = relevantChanges.filter(c => c.uri === uri);
+            await model.handleGitStatusUpdate(changes);
+        }
+    }
+
+    protected isEnabled(): boolean {
+        return this.preferences[DIRTYDIFF.ENABLED];
+    }
+
+    protected handlePreferenceChange(event: PreferenceChangeEvent<DirtyDiffConfiguration>) {
+        const { preferenceName, newValue } = event;
+        if (preferenceName === DIRTYDIFF.ENABLED) {
+            const enabled = !!newValue;
+            for (const model of this.models.values()) {
+                model.enabled = enabled;
+                model.update();
+            }
+        }
+    }
+
+}
+
+export interface DityDiffUpdate extends DirtyDiff {
+    readonly uri: string;
+}
+
+export class DirtyDiffModel implements Disposable {
+
+    editorCounter: number = 0;
+    enabled = true;
+
+    protected dirty = true;
+    protected staged: boolean;
+    protected previousContent: string[];
+    protected currentContent: string[];
+
+    protected readonly onDityDiffUpdateEmitter = new Emitter<DityDiffUpdate>();
```

Comment:
don't forget to dispose emitters

Reply:
it's disposed in `DirtyDiffModel.dispose`.